uefi/src/proto/ata/mod.rs

```diff
@@ -0,0 +1,323 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+//! ATA Protocols.
+
+use crate::mem::{AlignedBuffer, AlignmentError};
+use core::alloc::LayoutError;
+use core::marker::PhantomData;
+use core::ptr;
+use core::time::Duration;
+use uefi_raw::protocol::ata::{
+    AtaCommandBlock, AtaPassThruCommandPacket, AtaPassThruLength, AtaStatusBlock,
+};
+
+pub mod pass_thru;
+
+/// Represents the protocol for ATA Pass Thru command handling.
+///
+/// This type defines the protocols supported for passing commands through the ATA controller.
+pub use uefi_raw::protocol::ata::AtaPassThruCommandProtocol;
+
+/// Represents an ATA request built for execution on an ATA controller.
+#[derive(Debug)]
+pub struct AtaRequest<'a> {
+    io_align: u32,
+    acb: AtaCommandBlock,
+    packet: AtaPassThruCommandPacket,
+    in_data_buffer: Option<AlignedBuffer>,
+    out_data_buffer: Option<AlignedBuffer>,
+    asb: AlignedBuffer,
+    _phantom: PhantomData<&'a u8>,
+}
+
+/// Builder for creating and configuring an [`AtaRequest`].
+///
+/// This builder simplifies the creation of an [`AtaRequest`] by providing chainable methods for
+/// configuring fields like timeout, buffers, and ATA command details.
+#[derive(Debug)]
+pub struct AtaRequestBuilder<'a> {
+    req: AtaRequest<'a>,
+}
+
+impl<'a> AtaRequestBuilder<'a> {
+    /// Creates a new [`AtaRequestBuilder`] with the specified alignment, command, and protocol.
+    ///
+    /// # Parameters
+    /// - `io_align`: The I/O buffer alignment required for the ATA controller.
+    /// - `command`: The ATA command byte specifying the operation to execute.
+    /// - `protocol`: The protocol type for the command (e.g., DMA, UDMA, etc.).
+    ///
+    /// # Returns
+    /// `Result<Self, LayoutError>` indicating success or memory allocation failure.
+    ///
+    /// # Errors
+    /// This method can fail due to alignment or memory allocation issues.
+    fn new(
+        io_align: u32,
+        command: u8,
+        protocol: AtaPassThruCommandProtocol,
+    ) -> Result<Self, LayoutError> {
+        // status block has alignment requirements!
+        let mut asb =
+            AlignedBuffer::from_size_align(size_of::<AtaStatusBlock>(), io_align as usize)?;
```

Comment:
Nit: use [`usize_from_u32`](https://github.com/rust-osdev/uefi-rs/blob/cf4f328ccd9ec613c5a9d9d54fdeac58fd53cbaa/uefi/src/util.rs#L21) instead of `as` cast. (Eventually I hope to disable `as` casts via clippy lint, so trying to avoid introducing new ones)